### PR TITLE
Add test cases for NETFramework netstandard2 support

### LINKED ISSUE #1287: Add tests for .NET 4.6.1 support for .NET Standard 2.0
Project scenarios:
1.	SDK project building from CLI
2.	SDK project building from machine-wide MSBuild
a.	*also from VS
3.	Non-SDK project with PackageReferences
4.	Non-SDK project with project.json
5.	Non-SDK project with packages.config
6.	5 & 6 in VS 2015

Reference scenarios
1.	Hello world, no change
2.	Reference netstandard2-based-library and use a type in it, adds files to reference / output and compile succeeds.
3.	Reference old packages and netstandard2 library, no conflicts.
4.	Target net47 and use System.ValueTuple.  (It’s was moved from oob assembly to mscorlib, we have a different OOB assembly on net47 that type-forwards).

Future proofing scenarios:
1.	Add netstandard.dll to facades folder for target framework and targets shuts off injections.

Other:
- [x] Ensure retargeting experience works correctly in SDK and legacy projects. (test)
- [x] Ensure future framework that includes `netstandard2.0` support does not use these. (test)
- [x] Ensure binaries in support set are serviceable through packages. (test)

This is a piece of the work required for #1267, which tracks the overall work.

### PR DESCRIPTION
**Pull request by ericstj**
This adds test cases for validating inclusion of netstandard2 support libraries when targeting desktop.

Fixes #1287 

Scenarios not yet covered: non-SDK projects (PackageReference, project.json, packages.config)

In doing this work I discovered one issue which impacts the ValueTuple scenario: my exclusions were missing a trailing ')' causing us to select the wrong ValueTuple assembly (both were added and the first one was selected which was wrong for net47).

Note that for the net47/ValueTuple test to build correctly the machine requires the net47 targeting pack to be installed.  I'm not sure if this will be the case on all machines so we might want to guard that with a check for the targeting pack.

/cc @dsplaisted @nguerrera

### CODE REVIEW

**Comment by dsplaisted** [comment]
To fix the `All_props_and_targets_add_themselves_to_MSBuildAllTargets` test, make sure all of the new props / targets files that have been added have the following:

```xml
  <PropertyGroup>
    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
  </PropertyGroup>
```

**Comment by ericstj** [comment]
Yeah, I have a local fix for that as well as some other tests.  Hold off on review for a moment.

**Comment by ericstj** [comment]
This should be good now.  PTAL.

**Comment by ericstj** [comment]
Looking into the full framework failures.

**Comment by ericstj** [comment]
To fix the sequencing issue I'm going to further simplify the sequencing logic for the extensions _HandlePackageFileConflicts.  Previously I was careful to not set AfterTargets="ResolveNuGetPackageAssets" because that could cause log spew about an unknown target to appear in projects that didn't have the nuget targets.  That's no longer a concern since we're only installing to VS versions that have the target.  This lets me collapse the last two cases to be the same.

**Comment by ericstj** [comment]
Ok, this is ready to go.  @dsplaisted @nguerrera can one of you review please?

**Comment by dsplaisted** [comment]
@MattGertz for approval.  This PR includes tests and bugfixes to the .NET Framework support for .NET Standard.  Original template is [here](https://github.com/dotnet/sdk/pull/1279#issuecomment-305942903)

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+        [InlineData(true, ReferenceScenario.ProjectReference)]
+        [InlineData(true, ReferenceScenario.RawFileName)]
+        [InlineData(true, ReferenceScenario.HintPath)]
+        [InlineData(false, ReferenceScenario.ProjectReference)]
+        [InlineData(false, ReferenceScenario.RawFileName)]
+        [InlineData(false, ReferenceScenario.HintPath)]
+        public void It_includes_netstandard(bool isSdk, ReferenceScenario scenario)
+        {
+            var testAsset = _testAssetsManager
+                .CopyTestAsset(GetTemplateName(isSdk), identifier: scenario.ToString())
```
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs]: I think the identifier should include bot the value of `isSdk` as well as the `scenario` in order to use a unique path for each case of the theory.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopLibrary.cs [hunk 2]
```diff
@@ -91,6 +92,113 @@ public void Method1()
...
+            buildCommand
+                .Execute()
+                .Should()
+                .Pass();
+        }
+
+        [WindowsOnlyFact]
+        public void It_can_use_ValueTuple_and_exchange_the_type_with_a_NETStandard_library()
+        {
+            var referenceAssemblies = ToolLocationHelper.GetPathToDotNetFrameworkReferenceAssemblies(TargetDotNetFrameworkVersion.Version47);
```
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopLibrary.cs]: I don't think we'll get any automated coverage of this.  Are the .NET 4.7 reference assemblies part of the VS install?  They aren't on my machine.

**Review by dsplaisted — APPROVED** [review]

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs (hunk 1 above)
Comment (ericstj) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs]: They use different templates which get included in the path.  I was actually hitting max-path at one point which is what I shortened the test method names and tried to be brief in these identifiers.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopLibrary.cs (hunk 2 above)
Comment (ericstj) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopLibrary.cs]: I downloaded and installed it.  It's an optional component starting with 15.1.  Do you think we can install it on the CI machines?  Should be as simple as running https://www.microsoft.com/en-us/download/details.aspx?id=55168.

**Pull request merged**